internal/transformers/estransforms/optionalchain.go

```diff
@@ -11,7 +11,222 @@ type optionalChainTransformer struct {
+		return res
+	}
+	return ch.Factory().UpdateParenthesizedExpression(node, expr)
+}
+
+func (ch *optionalChainTransformer) visitPropertyOrElementAccessExpression(node *ast.Expression, captureThisArg bool, isDelete bool) *ast.Expression {
+	if node.Flags&ast.NodeFlagsOptionalChain != 0 {
+		// If `node` is an optional chain, then it is the outermost chain of an optional expression.
+		return ch.visitOptionalExpression(node.AsNode(), captureThisArg, isDelete)
+	}
+	expression := ch.Visitor().VisitNode(node.Expression())
+	// Debug.assertNotNode(expression, isSyntheticReference); // !!!
+
+	var thisArg *ast.Expression
+	if captureThisArg {
+		if !transformers.IsSimpleCopiableExpression(expression) {
+			thisArg = ch.Factory().NewTempVariable()
+			ch.EmitContext().AddVariableDeclaration(thisArg)
+			expression = ch.Factory().NewAssignmentExpression(thisArg, expression)
+		} else {
+			thisArg = expression
+		}
+	}
+
+	if node.Kind == ast.KindPropertyAccessExpression {
+		p := node.AsPropertyAccessExpression()
+		expression = ch.Factory().UpdatePropertyAccessExpression(p, expression, nil, ch.Visitor().VisitNode(p.Name()))
+	} else {
+		p := node.AsElementAccessExpression()
+		expression = ch.Factory().UpdateElementAccessExpression(p, expression, nil, ch.Visitor().VisitNode(p.AsElementAccessExpression().ArgumentExpression))
+	}
+
+	if thisArg != nil {
+		res := ch.Factory().NewSyntheticReferenceExpression(expression, thisArg)
+		ch.EmitContext().SetOriginal(res, node.AsNode())
+		return res
+	}
+	return expression
+}
+
+func (ch *optionalChainTransformer) visitDeleteExpression(node *ast.DeleteExpression) *ast.Node {
+	unwrapped := ast.SkipParentheses(node.Expression)
+	if unwrapped.Flags&ast.NodeFlagsOptionalChain != 0 {
+		return ch.visitNonOptionalExpression(node.Expression, false, true)
+	}
+	return ch.Visitor().VisitEachChild(node.AsNode())
+}
+
+func (ch *optionalChainTransformer) visitNonOptionalExpression(node *ast.Expression, captureThisArg bool, isDelete bool) *ast.Expression {
+	switch node.Kind {
+	case ast.KindParenthesizedExpression:
+		return ch.visitParenthesizedExpression(node.AsParenthesizedExpression(), captureThisArg, isDelete)
+	case ast.KindElementAccessExpression, ast.KindPropertyAccessExpression:
+		return ch.visitPropertyOrElementAccessExpression(node, captureThisArg, isDelete)
+	case ast.KindCallExpression:
+		return ch.visitCallExpression(node.AsCallExpression(), captureThisArg)
+	default:
+		return ch.Visitor().VisitNode(node.AsNode())
+	}
+}
+
+type flattenResult struct {
+	expression *ast.Expression
+	chain      []*ast.Node
+}
+
+func flattenChain(chain *ast.Node) flattenResult {
+	// Debug.assertNotNode(chain, isNonNullChain); // !!!
+	links := []*ast.Node{chain}
+	for !ast.IsTaggedTemplateExpression(chain) && chain.QuestionDotToken() == nil {
+		chain = ast.SkipPartiallyEmittedExpressions(chain.Expression())
+		// Debug.assertNotNode(chain, isNonNullChain); // !!!
+		links = append([]*ast.Node{chain}, links...)
+	}
```

Comment:
I know that this was the original but it seems so not-ideal - why didn't we just keep appending and reverse?


```suggestion
		links = append(links, chain)
	}
    slices.Reverse(links)
```